### ISSUE TITLE
Provide `.ipynb` file support out of the box

### ISSUE THREAD

**Issue by joyceerhl** [microsoft/vscode#128495]
Work items
- [x] Built-in extension that ships with VS Code and that can be disabled but not uninstalled
        - Support both web and desktop
        - Starter code https://github.com/joyceerhl/vscode-jupyter-notebook-serializer
- [ ] ~~Built-in support for creating a new Jupyter notebook via the New File... flow~~
- [x] `ms-toolsai.jupyter` cleanup
    - [x] Deregister our serializer https://github.com/microsoft/vscode-jupyter/pull/6778
    - [x] Delay native notebooks survey notification when users dismiss (eventually remove survey) https://github.com/microsoft/vscode-jupyter/pull/6767
    - [x] Do not update workbench.editorAssociations setting based on experiment (VS Code will continue to cache) https://github.com/microsoft/vscode-jupyter/pull/6768
- [ ] New Jupyter notebook getting started experience in VS Code
    - [x] Recommend Jupyter extension if missing renderers for output mimetypes
    - [x] Recommend kernel extensions in order to run code
    - [x] Fix existing notification to install Jupyter extension for ipynb files (shows only for json, not for notebook editor UI which is the default experience today)
    - [x] Syntax highlighting for ipynb json
- [x] Restore custom editor backup with ipynb serializer @rebornix
  - [ ] ~Remove custom editor contribution in Jupyter extension (after testing)~

Deferred
- [x] Jupyter and builtin ipynb extension to depend on a common npm package providing utilities for converting between VS Code notebook types and nbformat
    - [x] Reimplement create new and export without serializer https://github.com/microsoft/vscode-jupyter/pull/6756

**Comment by rebornix** [comment]
We would also want to support syntax highlighting `ipynb` when it's opened in a text editor. We would need a grammar which can be identical to JSON (but we can't add `ipynb` as a file ext option for JSON ext otherwise it will break save).

**Comment by joyceerhl** [comment]
Discussion items from meeting with @DonJayamanne and @roblourens:
- Supporting newer versions of `nbformat`
    -  Not currently necessary but in future this extension will need to be updated fairly regularly as `nbformat` ships updates
- Notebook and cell metadata
    - The current serializer adds metadata under a `custom` property
    - Some of this is done to support .NET Interactive scenarios, e.g. see https://github.com/microsoft/vscode-jupyter/pull/6196
    - No immediate plans to change this but if we do we should inform .NET Interactive team and others
- Compression of cell outputs on save
    - IIRC Don mentioned doing some compression of outputs on save rather than at the time that the outputs are updated, the built-in serializer will not do this

**Comment by rebornix** [comment]
re 

> Compression of cell outputs on save
> IIRC Don mentioned doing some compression of outputs on save rather than at the time that the outputs are updated, the built-in serializer will not do this

maybe it's related to terminal outputs which we currently compress in notebook controllers, which leads to performance issue. We may want to just do this on save.

**Comment by roblourens** [comment]
@rebornix what do you mean it will "break save" if we call ipynb files JSON?

**Comment by joyceerhl** [comment]
@roublourens if we call ipynb files JSON, when doing Save As on an untitled Jupyter notebook, the file picker attempts to save it as JSON by default and fails with an error.

**Comment by rebornix** [comment]
@joyceerhl already gave it a good summary. The issues tracking them are:

* No syntax highlighting for ipynb https://github.com/microsoft/vscode/issues/122589 and my original fix was adding `ipynb` to the JSON file extension list, which caused below issue
* https://github.com/microsoft/vscode/issues/124918

**Comment by davidanthoff** [comment]
Quick question: does this essentially mean that if a user has VS Code and the Julia extension installed, they would no longer even need the Jupyter extension to work with Jupyter notebook files?

**Comment by joyceerhl** [comment]
This means that you wouldn't need the Jupyter extension installed to open ipynb files. Renderers, Jupyter kernels, and other functionality like export are currently still provided through the Jupyter extension.

**Comment by davidanthoff** [comment]
We have the Julia notebook controller in the Julia extension now, so presumably the whole kernel functionality from the Jupyter extension wouldn't be needed at that point, and it would really only be renderers and export?

I hadn't even considered this scenario, but if we could get the Jupyter Julia Notebook scenario to work without even needing the Jupyter extension that of course would also be great! The less stuff users have to install, the better :)

**Comment by joyceerhl** [comment]
> the whole kernel functionality from the Jupyter extension wouldn't be needed at that point, and it would really only be renderers and export?

That's correct 😊

**Comment by davidanthoff** [comment]
Are there plans to move the renderers into core as well?

**Comment by joyceerhl** [comment]
Not currently because the renderers are very large.

**Comment by DonJayamanne** [comment]
> re there plans to move the renderers into core as well?

We're looking into  shiping the renderers as a seprate extension, see here https://github.com/microsoft/vscode-internalbacklog/blob/main/plan-master-copies/current/July%202021%20-%20jupyter.md#engineering

If this goes ahead, then Julia would not need Jupyter at all, just the renderer extension

**Comment by davidanthoff** [comment]
That would be fantastic! I think in that case we would presumably just require the renderer extension as a dep of our extension so that it is always installed if someone installs the Julia extension.

@DonJayamanne the link you posted is private, so I can't look at that.

**Comment by DonJayamanne** [comment]
Oops, sorry forgot about that.

> ust require the renderer extension as a dep of our extension so that it is always installed if someone installs the Julia extension.

Hopefully yes

**Comment by davidanthoff** [comment]
Just one other bit of feedback: the biggest problem for the Julia extension at the moment is that we can't save language and kernelspec info in Jupyter notebooks, see [here](https://github.com/julia-vscode/julia-vscode/issues/2256). This is blocking for us, and for now we have actually disabled the native notebook integration by default via a feature flag in the extension. We originally had hoped to ship the native notebook stuff on Wed as part of our Juliacon announcement, but what we're going to do instead now is ship it but have it disabled by default, and just tell people that if they are ok with notebooks not being saved properly, they can enable the Julia notebook controllers via config. Not great, but as far as I understand it the best we can do at the moment.

I don't understand how this scenario can be unblocked (i.e. whether we need something in VS Code or the Jupyter extension), but it would be fantastic if that could be a priority so that we can actually get the notebook stuff out to everyone :)

**Comment by joyceerhl** [comment]
@davidanthoff Based on https://github.com/julia-vscode/julia-vscode/pull/2271#discussion_r665912172 it sounds like the Julia extension would be unblocked if the [notebook metadata edit API](https://github.com/microsoft/vscode/issues/106744) is stabilized, since what you need is the ability to modify the `"language_info"` and `"kernel_info"` metadata properties. Is that accurate, and would the Julia extension need anything else?

**Comment by davidanthoff** [comment]
Yes, exactly! We essentially just need the ability to finish [this](https://github.com/julia-vscode/julia-vscode/blob/28ebc28d0b0cf3a9d5df280de658904827cedb88/src/notebook/notebookFeature.ts#L168) step.

I'll have a bunch of other feedback around kernel select, but as far as I can tell right now none of that is blocking, I think the metadata stuff is the only thing that I can think of right now that is a blocker on turning the switch to make the notebook functionality on by default.

**Comment by DonJayamanne** [comment]
Filed an issue that should unblock Julia (with a generic way to update metadata) https://github.com/microsoft/vscode/issues/129595

**Comment by davidanthoff** [comment]
Great, thanks!

**Comment by roblourens** [comment]
Closing, the only remaining task is the getting started experience which is covered by other issues
